tests/nixos/testsOnNixOS.nix

```diff
@@ -0,0 +1,72 @@
+{ lib, ... }:
+
+let
+  # FIXME (roberth) reference issue
+  inputDerivation = pkg: (pkg.overrideAttrs (o: {
+    disallowedReferences = [ ];
+  })).inputDerivation;
+
+in
+{
+  nodes.machine = { config, pkgs, ... }: {
+
+    virtualisation.writableStore = true;
+    system.extraDependencies = [
+      (inputDerivation config.nix.package)
+    ];
+
+    nix.package = (builtins.getFlake "git+file:///home/user/h/nix?rev=a1c7cb8fb6c6511daf4884e140aba8ae1788cd84").packages.x86_64-linux.default;
+    nix.settings.substituters = lib.mkForce [];
+
+    nixpkgs.overlays = [
+      (final: prev: {
+        thisNix = prev.nix;
+        nix = config.nix.package;
+      })
+    ];
```

Comment:
The purpose of the `nix.package` and overlay hacks here is to not rebuild the `nix` derivation every time the test suite is changed during local development. This hack should be moved to a convenience attribute and not run on CI. It should use `HEAD` instead of a specific hash.
Alternatively, we could have a mode that adds an impurely built (devshell) nix into the store somehow, or we could build Nix in such a way that it does not depend on its test suite.
 - Possibly helpful is #7662, to recover the testedness guarantee (or work around it by adding an indirection that runs both types of test)